enterprise/reporting/utils.go

```diff
@@ -0,0 +1,103 @@
+package reporting
+
+import (
+	"encoding/json"
+	"sort"
+	"strings"
+
+	"github.com/rudderlabs/rudder-go-kit/config"
+	"github.com/rudderlabs/rudder-server/enterprise/reporting/event_sampler"
+	"github.com/rudderlabs/rudder-server/utils/types"
+)
+
+func floorFactor(intervalMs int64) int64 {
+	factors := []int64{1, 2, 3, 4, 5, 6, 10, 12, 15, 20, 30, 60}
+
+	// Find the smallest index where factors[i] >= intervalMs
+	index := sort.Search(len(factors), func(i int) bool {
+		return factors[i] >= intervalMs
+	})
+
+	// If index is 0, intervalMs is smaller than the smallest factor
+	if index == 0 {
+		return factors[0]
+	}
+
+	// If factors[index] == intervalMs, return it directly
+	if index < len(factors) && factors[index] == intervalMs {
+		return factors[index]
+	}
+
+	// Otherwise, return the previous factor
+	return factors[index-1]
+}
+
+func getAggregationBucketMinute(timeMs, intervalMs int64) (int64, int64) {
+	// If interval is not a factor of 60, then the bucket start will not be aligned to hour start
+	// For example, if intervalMs is 7, and timeMs is 28891085 (6:05) then the bucket start will be 28891079 (5:59)
+	// and current bucket will contain the data of 2 different hourly buckets, which is should not have happened.
+	// To avoid this, we round the intervalMs to the nearest factor of 60.
+	if intervalMs <= 0 || 60%intervalMs != 0 {
+		intervalMs = floorFactor(intervalMs)
+	}
+	bucketStart := timeMs - (timeMs % intervalMs)
+	bucketEnd := bucketStart + intervalMs
+	return bucketStart, bucketEnd
+}
+
+func transformMetricWithEventSampling(metric types.PUReportedMetric, reportedAt int64, eventSampler event_sampler.EventSampler, eventSamplingDuration int64) (types.PUReportedMetric, error) {
+	if eventSampler == nil {
+		return metric, nil
+	}
+
+	isValidSampleEvent := metric.StatusDetail.SampleEvent != nil && string(metric.StatusDetail.SampleEvent) != "{}"
+
+	if isValidSampleEvent {
```

Comment:
@itsmihir @vamsikrishnakandi We need a fix here similar to [this pr](https://github.com/rudderlabs/rudder-server/pull/5360)

```suggestion
	isValidSample := (metric.StatusDetail.SampleEvent != nil && string(metric.StatusDetail.SampleEvent) != "{}") ||
		metric.StatusDetail.SampleResponse != ""

	if isValidSample {
```